core/blockchain_reader.go

```diff
@@ -49,6 +49,22 @@ func (bc *BlockChain) CurrentFastBlock() *types.Block {
 	return bc.currentFastBlock.Load().(*types.Block)
 }
 
+// CurrentTerminalHeader retrieves the current terminal header of the canonical
+// chain. The block is retrieved from the blockchain's internal cache.
+func (bc *BlockChain) CurrentTerminalHeader() *types.Header {
+	if head := rawdb.ReadTerminalBlockHash(bc.db); head != (common.Hash{}) {
+		if header := bc.GetHeaderByHash(head); header != nil {
+			return header
+		}
+	}
+	return nil
+}
+
+// SetTerminalBlock sets the current terminal block.
+func (bc *BlockChain) SetTerminalBlock(header *types.Header) {
+	rawdb.WriteTerminalBlockHash(bc.db, header.Hash())
+}
```

Comment:
This method only writes the `TerminalBlockHash`.
I think it's semantically a bit wrong. It should either
-  Be called SetTerminalBlockHash and take a hash
- OR write both the hash, and the header itself


Just my 2cents